jfr-connection/src/main/java/io/opentelemetry/contrib/jfr/connection/FlightRecorderConnection.java

```diff
@@ -0,0 +1,309 @@
+   *     this class.
+   */
+  public long startRecording(
+      RecordingOptions recordingOptions, RecordingConfiguration recordingConfiguration)
+      throws IOException, JfrStreamingException {
+
+    try {
+      Object[] args = new Object[] {};
+      String[] argTypes = new String[] {};
+      long id = (long) mBeanServerConnection.invoke(objectName, "newRecording", args, argTypes);
+
+      if (recordingConfiguration != null) {
+        setConfiguration(recordingConfiguration, id);
+      }
+
+      if (recordingOptions != null) {
+        setOptions(recordingOptions, id);
+      }
+
+      args = new Object[] {id};
+      argTypes = new String[] {long.class.getName()};
+      mBeanServerConnection.invoke(objectName, "startRecording", args, argTypes);
+
+      return id;
+    } catch (OpenDataException
+        | InstanceNotFoundException
+        | MBeanException
+        | ReflectionException e) {
+      // In theory, we should never get these.
+      throw new JfrStreamingException(e.getMessage(), e);
+    }
+  }
+
+  private void setOptions(RecordingOptions recordingOptions, long id)
+      throws OpenDataException, InstanceNotFoundException, MBeanException, ReflectionException,
+          IOException {
+    Map<String, String> options = recordingOptions.getRecordingOptions();
+    if (options != null && !options.isEmpty()) {
+      TabularData recordingOptionsParam = OpenDataUtils.makeOpenData(options);
+      Object[] args = new Object[] {id, recordingOptionsParam};
+      String[] argTypes = new String[] {long.class.getName(), TabularData.class.getName()};
+      mBeanServerConnection.invoke(objectName, "setRecordingOptions", args, argTypes);
+    }
+  }
+
+  private void setConfiguration(RecordingConfiguration recordingConfiguration, long id)
+      throws OpenDataException, InstanceNotFoundException, MBeanException, ReflectionException,
+          IOException {
+    recordingConfiguration.invokeSetConfiguration(id, mBeanServerConnection, objectName);
+  }
+
+  /**
+   * Stop a recording. This method is called from the {@link Recording#stop()} method.
+   *
+   * @param id The id of the recording.
+   * @throws IOException A communication problem occurred when talking to the MBean server.
+   * @throws JfrStreamingException Wraps an {@code javax.management.InstanceNotFoundException}, a
+   *     {@code javax.management.MBeanException} or a {@code javax.management.ReflectionException}
+   *     and indicates an issue with the FlightRecorderMXBean in the JVM.
+   */
+  public void stopRecording(long id) throws IOException, JfrStreamingException {
+    try {
+      Object[] args = new Object[] {id};
+      String[] argTypes = new String[] {long.class.getName()};
+      mBeanServerConnection.invoke(objectName, "stopRecording", args, argTypes);
+    } catch (InstanceNotFoundException | MBeanException | ReflectionException e) {
+      throw new JfrStreamingException(e.getMessage(), e);
```

Comment:
Done. I added a canonicalJfrConnectionException method in FlightRecorderConnection and use it pretty much everywhere. I toyed with adding a constructor to JfrConnectionException to create a canonical message, but decided against it because the canonical message is just one possible use of the message field.